extensions/wikia/CreateNewWiki/CreateNewWikiController.class.php

```diff
@@ -228,132 +241,137 @@ public function CreateWiki() {
+			wfProfileOut(__METHOD__);
+			return;
+		}
 
-				$this->status = 'backenderror';
-				$this->statusMsg = wfMessage( 'cnw-error-general' )->parse();
-				$this->statusHeader = wfMessage( 'cnw-error-general-heading' )->escaped();
+		$categories = isset($params['wCategories']) ? $params['wCategories'] : array();
 
-				$this->errClass = get_class( $ex );
-				$this->errCode = $ex->getCode();
-				$this->errMessage = $ex->getMessage();
+		$createWiki = new CreateWiki($params['wName'], $params['wDomain'], $params['wLanguage'], $params['wVertical'], $categories);
 
-				$this->error('CreateWiki: failed to create new wiki', [
+		try {
+			$createWiki->create();
+		}
+		catch(Exception $ex) {
+			$error_code = $ex->getCode();
+			$this->error(
+				__METHOD__ . ": backend failed to process the request: " . $ex->getMessage(),
+				[
 					'code' => $error_code,
 					'params' => $params,
 					'exception' => $ex
-				]);
-
-				wfProfileOut( __METHOD__);
-				return;
-			}
+				]
+			);
+			$this->response->setCode( 500 );
+			$this->response->setVal( self::STATUS_FIELD, self::STATUS_BACKEND_ERROR );
+			$this->response->setVal( self::STATUS_MSG_FIELD, wfMessage( 'cnw-error-general' )->parse() );
+			$this->response->setVal( self::STATUS_HEADER_FIELD, wfMessage( 'cnw-error-general-heading' )->escaped() );
+			$this->response->setVal( self::ERROR_CLASS_FIELD, get_class( $ex ) );
+			$this->response->setVal( self::ERROR_CODE_FIELD, $ex->getCode() );
+			$this->response->setVal( self::ERROR_MESSAGE_FIELD, $ex->getMessage() );
+			wfProfileOut( __METHOD__);
+			return;
+		}
 
-			$cityId = $createWiki->getWikiInfo('city_id');
+		$cityId = $createWiki->getWikiInfo('city_id');
 
-			if ( isset($params['wAllAges']) && !empty( $params['wAllAges'] ) ) {
-				WikiFactory::setVarByName( self::WF_WDAC_REVIEW_FLAG_NAME, $cityId, true, __METHOD__ );
-			}
-			$this->status = 'ok';
-			$this->siteName = $createWiki->getWikiInfo('sitename');
-			$this->cityId = $cityId;
-			$finishCreateTitle = GlobalTitle::newFromText("FinishCreate", NS_SPECIAL, $cityId);
-			$this->finishCreateUrl = empty($wgDevelDomains) ? $finishCreateTitle->getFullURL() : str_replace('.wikia.com', '.'.$wgDevelDomains[0], $finishCreateTitle->getFullURL());
-
-			$this->info('CreateWiki: completed', [
-				'city_id' => $cityId,
-				'params' => $params,
-			]);
+		if ( isset($params['wAllAges']) && !empty( $params['wAllAges'] ) ) {
+			WikiFactory::setVarByName( self::WF_WDAC_REVIEW_FLAG_NAME, $cityId, true, __METHOD__ );
 		}
 
+		$this->response->setVal( self::STATUS_FIELD, self::STATUS_OK );
+		$this->response->setVal( self::SITE_NAME_FIELD, $createWiki->getWikiInfo('sitename') );
+		$this->response->setVal( self::CITY_ID_FIELD, $cityId );
+		$finishCreateTitle = GlobalTitle::newFromText("FinishCreate", NS_SPECIAL, $cityId);
+		$finishCreateUrl = empty( $wgDevelDomains ) ? $finishCreateTitle->getFullURL() : str_replace( '.wikia.com', '.'.$wgDevelDomains[0], $finishCreateTitle->getFullURL() );
```

Comment:
can we move this to a separate function? also, we could cache `$finishCreateTitle->getFullURL()` into var, not to call this function twice.